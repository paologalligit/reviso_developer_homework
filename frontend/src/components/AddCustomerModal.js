/* eslint-disable max-len */
/* eslint-disable react/prop-types */
import React from 'react';
import {
  Form, Input, Button, Modal,
} from 'semantic-ui-react';
import { withFormik } from 'formik';
import { compose, graphql } from 'react-apollo';
import isDecimal from 'validator/lib/isDecimal';

import getCustomersPerUser from '../graphql/query/customer';
import normalizeErrors from '../normalizeErrors';
import createCustomerMutation from '../graphql/mutation/customer';

// userId,
// console.log('errors: ', touched);
const AddCustomerModal = ({
  open,
  onClose,
  // userId,
  values,
  handleChange,
  handleBlur,
  handleSubmit,
  isSubmitting,
  resetForm,
  errors,
  touched,
}) => (
  <Modal
    open={open}
    onClose={(e) => {
      resetForm();
      onClose(e);
    }}
  >
    <Modal.Header>Add Customer</Modal.Header>
    <Modal.Content>
      <Form>
        <Form.Group unstackable widths={3}>
          <Form.Field error={!!errors.name}>
            <Input
              value={values.name}
              onChange={handleChange}
              onBlur={handleBlur}
              name="name"
              fluid
              placeholder="Name"
            />
            {touched.name && errors.name ? errors.name[0] : null}
          </Form.Field>
          <Form.Field error={!!errors.surname}>
            <Input
              value={values.surname}
              onChange={handleChange}
              onBlur={handleBlur}
              name="surname"
              fluid
              placeholder="Surname"
            />
            {touched.surname && errors.surname ? errors.surname[0] : null}
          </Form.Field>
          <Form.Field error={!!errors.email}>
            <Input
              value={values.email}
              onChange={handleChange}
              onBlur={handleBlur}
              name="email"
              fluid
              placeholder="Email"
            />
            {touched.email && errors.email ? errors.email[0] : null}
          </Form.Field>
        </Form.Group>

        <Form.Group unstackable widths={3}>
          <Form.Field error={!!errors.country}>
            <Input
              value={values.country}
              onChange={handleChange}
              onBlur={handleBlur}
              name="country"
              fluid
              placeholder="Country"
            />
            {touched.country && errors.country ? errors.country[0] : null}
          </Form.Field>
          <Form.Field error={!!errors.city}>
            <Input
              value={values.city}
              onChange={handleChange}
              onBlur={handleBlur}
              name="city"
              fluid
              placeholder="City"
            />
            {touched.city && errors.city ? errors.city[0] : null}
          </Form.Field>
          <Form.Field error={!!errors.postal}>
            <Input
              value={values.postal}
              onChange={handleChange}
              onBlur={handleBlur}
              name="postal"
              placeholder="Postal"
            />
            {touched.postal && errors.postal ? errors.postal[0] : null}
          </Form.Field>
        </Form.Group>

        <Form.Group unstackable widths={2}>
          <Form.Field error={!!errors.address}>
            <Input
              value={values.address}
              onChange={handleChange}
              onBlur={handleBlur}
              name="address"
              fluid
              placeholder="Address"
            />
            {touched.address && errors.address ? errors.address[0] : null}
          </Form.Field>
          <Form.Field>
            <Input
              value={values.specialization}
              onChange={handleChange}
              onBlur={handleBlur}
              name="specialization"
              fluid
              placeholder="Specialization"
            />
          </Form.Field>
        </Form.Group>

        <Form.Group widths="equal">
          <Button
            disabled={isSubmitting}
            fluid
            onClick={(e) => {
              resetForm();
              onClose(e);
            }}
          >
              Cancel
          </Button>
          <Button disabled={isSubmitting} onClick={handleSubmit} fluid>
              Create Customer
          </Button>
        </Form.Group>
      </Form>
    </Modal.Content>
  </Modal>
);

export default compose(
  graphql(createCustomerMutation),
  withFormik({
    mapPropsToValues: ({ userId }) => ({
      userId,
      name: '',
      surname: '',
      country: '',
      email: '',
      city: '',
      specialization: '',
      address: '',
      postal: '',
    }),
    handleSubmit: async (
      values,
      { props: { onClose, userId, mutate }, setSubmitting, setErrors },
    ) => {
      if (isDecimal(values.postal)) {
        const response = await mutate({
          variables: {
            name: values.name,
            surname: values.surname,
            country: values.country,
            email: values.email,
            city: values.city,
            specialization: values.specialization,
            address: values.address,
            postal: parseInt(values.postal, 10),
            user_id: userId,
          },
          optimisticResponse: {
            __typename: 'Mutation',
            registerCustomer: {
              __typename: 'Mutation',
              ok: true,
              errors: null,
              customer: {
                __typename: 'RegisterCustomer',
                id: -1,
                name: values.name,
                surname: values.surname,
                country: values.country,
                email: values.email,
                city: values.city,
                specialization: values.specialization,
                address: values.address,
                postal: parseInt(values.postal, 10),
                user_id: userId,
              },
            },
          },
          update: (store, { data: { registerCustomer } }) => {
            const { ok, customer } = registerCustomer;
            if (!ok) {
              return;
            }

            const data = store.readQuery({
              query: getCustomersPerUser,
              variables: { user_id: values.userId },
            });

            data.getCustomersPerUser.push(customer);

            store.writeQuery({
              query: getCustomersPerUser,
              variables: { user_id: values.userId },
              data,
            });
          },
        });

        const { ok, errors } = response.data.registerCustomer;
        if (ok) {
          onClose();
        } else {
          setErrors(normalizeErrors(errors));
        }
      } else {
        setErrors({
          postal: ['Postal number required'],
        });
      }

      setSubmitting(false);
    },
  }),
)(AddCustomerModal);
