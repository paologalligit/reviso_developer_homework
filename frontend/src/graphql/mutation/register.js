import gql from 'graphql-tag';

export default gql`
  mutation(
    $name: String!
    $surname: String!
    $username: String!
    $email: String!
    $password: String!
    $birthdate: String!
    $country: String!
    $city: String!
    $address: String!
    $postal: Int!
  ) {
    registerUser(
      name: $name
      surname: $surname
      username: $username
      email: $email
      password: $password
      birthdate: $birthdate
      country: $country
      city: $city
      address: $address
      postal: $postal
    ) {
      ok
      errors {
        path
        message
      }
    }
  }
`;
