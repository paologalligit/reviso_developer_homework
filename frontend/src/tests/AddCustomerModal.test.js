import 'jsdom-global/register';
import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow, render, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

import AddCustomerModal from '../components/AddCustomerModal';
import createCustomerMutation from '../graphql/mutation/customer';

describe('add customer modal rendering', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <MockedProvider>
        <AddCustomerModal />
      </MockedProvider>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('add customer modal interaction', () => {
  const correctMock = [
    {
      request: {
        query: createCustomerMutation,
        variables: {
          name: 'Biello',
          surname: 'Imployr',
          email: 'ymplo@email.com',
          country: 'Italy',
          city: 'Cibby',
          address: 'Downson Street',
          postal: 90887,
          specialization: 'Pollo',
          user_id: 1,
        },
      },
      result: {
        data: {
          registerCustomer: {
            ok: true,
            customer: {
              id: 1,
              name: 'Biello',
              surname: 'Imployr',
              email: 'ymplo@email.com',
              country: 'Italy',
              city: 'Cibby',
              address: 'Downson Street',
              postal: 90887,
              specialization: 'Pollo',
              user_id: 1,
            },
            errors: null,
          },
        },
      },
    },
  ];
  const onCloseMock = jest.fn();

  it('register a new customer', () => {
    const wrapper = mount(
      <MockedProvider mocks={correctMock} addTypename={false}>
        <AddCustomerModal open onClose={onCloseMock} />
      </MockedProvider>,
    );

    const modal = wrapper.find('Modal');
    // console.log('the modal: ', modal.debug())
    const modalProps = modal.props();
    /*
    const formsInput = modal.find('Input');
    // console.log(formsInput.debug());

    const nameIn = formsInput.at(0);
    const surnameIn = formsInput.at(1);
    const emailIn = formsInput.at(2);
    const countryIn = formsInput.at(3);
    const cityIn = formsInput.at(4);
    const postalIn = formsInput.at(5);
    const addressIn = formsInput.at(6);
    const specIn = formsInput.at(7);
    */
    // .simulate('change', {
    //   persist: () => {},
    //   target: {
    //     name: 'name',
    //     value: 'Biello',
    //   },
    // });

    /*
    wrapper.find('form').simulate('submit');

    console.log(
      wrapper
        .find('Input')
        .at(0)
        .props().value,
    );

    const newValue = wrapper
      .find(AddCustomerModal)
      .dive()
      .find('input')
      .at(0)
      .props().value;

    console.log('the value:_ ', newValue);
    */
    modalProps.onClose();
  });
});
