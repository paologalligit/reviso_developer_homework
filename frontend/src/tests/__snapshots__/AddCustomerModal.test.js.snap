// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add customer modal rendering renders correctly 1`] = `null`;
